### PR TITLE
FAPI 2.0 security profile - not allow an authorization request whose parameters were not included in Request Object pushed to PAR request

### LINKED ISSUE #20710: FAPI 2.0 security profile - not allow an authorization request whose parameters were not included in Request Object pushed to PAR request
### Description

According to [FAPI 2.0 security profile](https://openid.bitbucket.io/fapi/fapi-2_0-security-profile.html), all necessary request parameters need to be included  in an authorization request directly to the PAR endpoint. Therefore, if a client sends an authorization request to Authorization endpoint, but the request violates the point above, an authorization server needs to return an error response.

https://github.com/keycloak/keycloak/issues/20623 resolved it. However, it does not treat request object pushed to PAR endpoint.
When a request object is pushed to PAR endpoint, parsing the request object is needed.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by tnorimat**
closes #20710

### CODE REVIEW

File: services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureParContentsExecutor.java [hunk 1]
```diff
@@ -76,17 +85,66 @@ private void checkValidParContents(PreAuthorizationRequestContext preAuthorizati
...
+        for (String param : requestParametersFromPAR.keySet()) {
+            if (OIDCLoginProtocol.REQUEST_PARAM.equals(param)) continue;
+            parRetrievedRequest.add(param);
+        }
+        // from parsed PAR request parameters
+        if (request.getClientId() != null) parRetrievedRequest.add(OIDCLoginProtocol.CLIENT_ID_PARAM);
+        if (request.getResponseType() != null) parRetrievedRequest.add(OIDCLoginProtocol.RESPONSE_TYPE_PARAM);
+        if (request.getResponseMode() != null) parRetrievedRequest.add(OIDCLoginProtocol.RESPONSE_MODE_PARAM);
+        if (request.getRedirectUriParam() != null) parRetrievedRequest.add(OIDCLoginProtocol.REDIRECT_URI_PARAM);
+        if (request.getState() != null) parRetrievedRequest.add(OIDCLoginProtocol.STATE_PARAM);
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureParContentsExecutor.java:128]: It seems there are quite a lot of parameters here and I afraid a bit that when we introduce another OIDC parameter, we have 2 places to update and we may easily forget one of them.

I wonder about somehow re-use some parts of the code with `AuthorizationEndpoint.updateAuthenticationSession()`, which already lists some parameters in a bit similar way.

For example we can add this method on `AuthorizationEndpoint` (or maybe some dedicated util class):
```
public static performActionOnParameters(AuthorizationEndpointRequest request, BiConsumer<String, String> paramAction) {
    paramAction.accept(OIDCLoginProtocol.STATE_PARAM, request.getState());
    paramAction.accept(OIDCLoginProtocol.NONCE_PARAM, request.getNonce());
    ... Other parameters useful in both SecureParContentsExecutor and AuthorizationEndpoint.updateAuthenticationSession() ...

}
```

Then in `AuthorizationEndpoint.updateAuthenticationSession()`, some part with repeated parameters can be replaced with something like:
```
performActionOnParameters(request, (paramName, paramValue) -> if (paramValue != null) authenticationSession.setClientNote(paramName, paramValue));
```

When in `SecureParContentsExecutor` something like:
```
AuthorizationEndpoint.performActionOnParameters(request, (paramName, paramValue) -> if (paramValue != null) parRetrievedRequest.add(paramName);
```

This probably cannot be used for all the parameters (as some are still doing something different in `AuthorizationEndpoint.updateAuthenticationSession()` than `authenticationSession.setClientNote`, but hopefully for majority of them.

WDYT?

**Review by mposolda — COMMENTED** [review]
@tnorimat Added one comment inline. Could you please check it?

File: services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureParContentsExecutor.java (hunk 1 above)
Comment (tnorimat) [reply services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureParContentsExecutor.java:128]: @mposolda Yes, it is a good idea. I would like to incorporate it.

I investigated the following point about request parameters:
- [1] comparing request parametes that are treated as you mentiond in  `AuthorizationEndpoint.updateAuthenticationSession()` and request parameters included in `AuthzEndpointRequestParser.KNOWN_REQ_PARAMS`
- [2] comparing request parametes that are treated as you mentiond in  `SecureParContentsExecutor.getParRetrievedRequestParameters()` and request parameters included in `AuthzEndpointRequestParser.KNOWN_REQ_PARAMS`
- [3] comparing request parametes that are treated as you mentiond in  `AuthorizationEndpoint.updateAuthenticationSession()` and request parameters that are treated as you mentiond in  `SecureParContentsExecutor.getParRetrievedRequestParameters()` 

As for [1], `AuthorizationEndpoint.updateAuthenticationSession()` only includes
-  OAuth2Constants.DISPLAY
-  LocaleSelectorProvider.CLIENT_REQUEST_LOCALE

while `AuthorizationEndpoint.updateAuthenticationSession()` does not include
-  OIDCLoginProtocol.CLIENT_ID_PARAM
-  OIDCLoginProtocol.REDIRECT_URI_PARAM
-  OIDCLoginProtocol.REQUEST_PARAM
-  OIDCLoginProtocol.REQUEST_URI_PARAM
-  OIDCLoginProtocol.RESPONSE_TYPE_PARAM
-  OIDCLoginProtocol.UI_LOCALES_PARAM

As for [2], `SecureParContentsExecutor.getParRetrievedRequestParameters()` only includes
-  OAuth2Constants.DISPLAY

while `SecureParContentsExecutor.getParRetrievedRequestParameters()` does not include
-  OIDCLoginProtocol.REQUEST_PARAM
-  OIDCLoginProtocol.REQUEST_URI_PARAM

however, the lack of them are intentional so no problem.

As for [3], both `AuthorizationEndpoint.updateAuthenticationSession()`  and `SecureParContentsExecutor.getParRetrievedRequestParameters()` include
- OIDCLoginProtocol.ACR_PARAM
- OIDCLoginProtocol.CLAIMS_PARAM
- OIDCLoginProtocol.CODE_CHALLENGE_METHOD_PARAM
- OIDCLoginProtocol.CODE_CHALLENGE_PARAM
- OIDCLoginProtocol.LOGIN_HINT_PARAM
- OIDCLoginProtocol.MAX_AGE_PARAM
- OIDCLoginProtocol.NONCE_PARAM
- OIDCLoginProtocol.PROMPT_PARAM
- OIDCLoginProtocol.RESPONSE_MODE_PARAM
- OIDCLoginProtocol.SCOPE_PARAM
-  OIDCLoginProtocol.STATE_PARAM

Therefore, I will treat parameters in [3] as you suggested. Other parameters are treated by each `AuthorizationEndpoint.updateAuthenticationSession()` and  `SecureParContentsExecutor.getParRetrievedRequestParameters()` in as usual.

**Comment by tnorimat** [comment]
@mposolda I revised the PR as you suggested. It passed all CI tests. Could you check it?

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks!

**Pull request merged**